plugins/superio/fu-superio-it55-device.c

```diff
@@ -0,0 +1,680 @@
+/*
+ * Copyright (C) 2021, TUXEDO Computers GmbH
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include <fwupdplugin.h>
+
+#include "fu-superio-common.h"
+#include "fu-superio-it55-device.h"
+
+/* ROM of IT5570 consists of 64KB blocks. Blocks can be further subdivided in
+ * 256-byte chunks, which is especially visible when erasing the ROM. This is
+ * because in case of erasure offset within a block is specified in chunks (even
+ * though erasure is done one kilobyte at a time).
+ *
+ * Accessing ROM requires entering a special mode, which should be always left
+ * to restore normal operation of EC (handling of buttons, keyboard, etc.). */
+
+#define SIO_CMD_EC_WRITE_BLOCK		0x02
+#define SIO_CMD_EC_READ_BLOCK		0x03
+#define SIO_CMD_EC_ERASE_KBYTE		0x05
+#define SIO_CMD_EC_WRITE_1ST_KBYTE	0x06
+#define SIO_CMD_EC_ID			0xF0
+#define EC_ROM_ACCESS_ON_1		0xDE
+#define EC_ROM_ACCESS_ON_2		0xDC
+#define EC_ROM_ACCESS_OFF		0xFE
+
+#define BLOCK_SIZE			0x10000
+#define CHUNK_SIZE			0x100
+#define CHUNKS_IN_KBYTE			0x4
+#define CHUNKS_IN_BLOCK			0x100
+
+#define MAX_FLASHING_ATTEMPTS		5
+
+typedef enum {
+	AUTOLOAD_NO_ACTION,
+	AUTOLOAD_DISABLE,
+	AUTOLOAD_SET_ON,
+	AUTOLOAD_SET_OFF,
+} AutoloadAction;
+
+struct _FuEcIt55Device {
+	FuSuperioDevice		 parent_instance;
+	gchar			*prj_name;
+	AutoloadAction		 autoload_action;
+};
+
+G_DEFINE_TYPE (FuEcIt55Device, fu_superio_it55_device, FU_TYPE_SUPERIO_DEVICE)
+
+
+static void
+fu_superio_it55_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuEcIt55Device *self = FU_SUPERIO_IT55_DEVICE (device);
+
+	/* FuSuperioDevice->to_string */
+	FU_DEVICE_CLASS (fu_superio_it55_device_parent_class)->to_string (device, idt, str);
+
+	fu_common_string_append_kx (str, idt, "AutoloadAction", self->autoload_action);
+}
+
+static gboolean
+fu_superio_it55_device_ec_check (FuSuperioDevice *self, GError **error)
+{
+	guint8 tmp = 0;
+	gboolean is_it5570;
+
+	/* enter ROM access mode */
+	if (!fu_superio_device_ec_write_cmd (self, EC_ROM_ACCESS_ON_1, error) ||
+	    !fu_superio_device_ec_write_cmd (self, EC_ROM_ACCESS_ON_2, error))
+		goto error;
+
+	if (!fu_superio_device_ec_write_cmd (self, SIO_CMD_EC_ID, error))
+		goto error;
+	if (!fu_superio_device_ec_read_data (self, &tmp, error))
+		goto error;
+	is_it5570 = (tmp != 0x00 && tmp != 0xff);
+
+	/* leave ROM access mode */
+	if (!fu_superio_device_ec_write_cmd (self, EC_ROM_ACCESS_OFF, error))
+		return FALSE;
```

Comment:
shouldn't this be set after you've checked if it's the right part?  If it's not the right part, you're shooting in the dark

Reply:
I *think* it's okay, as we call the parent setup, which calls `fu_superio_device_check_id()` before the subclassed setup.

Reply:
> shouldn't this be set after you've checked if it's the right part? If it's not the right part, you're shooting in the dark

@superm1 `0xf0` requires entering this mode to succeed, so there isn't many options. Apparently this mode is common across some ECs (maybe for all `IT55*`, but only `IT5570` does 1KB erasure and requires weird separate write of the first kilobyte).

> I _think_ it's okay, as we call the parent setup, which calls `fu_superio_device_check_id()` before the subclassed setup.

@hughsie The check in `fu_superio_device_check_id()` is disabled because it uses `->port`, which isn't defined for this EC.

Reply:
buuuut, are we not allow-listing the models with the hwids in the quirk file anyway?

Reply:
Yes, correct. So the commands should be applicable for the EC as long as information in the quirks is accurate.

Reply:
If we're allowlisting all the machines by quirks the only value for it is if someone tries to add a machine to quirk list to see "if it works". In that regard I'd say this check unnecessary.

Reply:
Dropped the check.